### DISCUSSION TITLE
Update Rhino library

### DISCUSSION THREAD

**Discussion by skbeh** [PurpurMC/Purpur#1277] [__Completed__] [answered]
[Rhino 1.7.14](https://github.com/mozilla/rhino/releases/tag/Rhino1_7_14_Release) has published for a while, but Purpur is still using a [very outdated version](https://github.com/PurpurMC/Purpur/blob/9940dcf74be5d871c452590943b4e269b3f3b666/patches/server/0003-Rebrand.patch#L27).

Also, [this](https://github.com/PurpurMC/Purpur/blob/15a76c78b19bb93a5d788ac2dd2da6408ae6ab51/patches/server/0009-Configurable-entity-base-attributes.patch#L1058) should be

```java
value = (double) scriptEngine.eval("'use strict'; { const size = " + size + "; " + equation.get() + "; }");
```

to reduce js overhead and prevent side effect.

By the way, it would be better to avoid embedding a js engine just to compute health and damage.

**Comment by granny** [comment, marked as answer]
> [Rhino 1.7.14](https://github.com/mozilla/rhino/releases/tag/Rhino1_7_14_Release) has published for a while, but Purpur is still using a [very outdated version](https://github.com/PurpurMC/Purpur/blob/9940dcf74be5d871c452590943b4e269b3f3b666/patches/server/0003-Rebrand.patch#L27).

Appreciate the heads up on the outdated Rhino version. Will look into updating that when I have time.

> Also, [this](https://github.com/PurpurMC/Purpur/blob/15a76c78b19bb93a5d788ac2dd2da6408ae6ab51/patches/server/0009-Configurable-entity-base-attributes.patch#L1058) should be
> 
> ```java
> value = (double) scriptEngine.eval("'use strict'; { const size = " + size + "; " + equation.get() + "; }");
> ```
> 
> to reduce js overhead and prevent side effect.

Will look into that as well.

> By the way, it would be better to avoid embedding a js engine just to compute health and damage.

Currently it's used to allow the end-user to define their own equation for calculating a slime's health and damage (and for any other equations a server owner would like to modify easily). By default a slime's max health is `size^2`. With our implementation, the end-user can choose to make the slime's max health using whatever equation they'd like. What's an alternative you have in mind that doesn't over-complicate the config file for a server owner, which would allow us to avoid embedding a js engine?

The only reason we include rhino.js is because oracle removed nashorn. Otherwise we'd use that instead.

**Reply by skbeh** [reply]
It seems that the performance of JShell is bery bad, so it may not be a good alternative.

**Reply by granny** [reply]
> It seems that the performance of JShell is bery bad, so it may not be a good alternative.

Interesting. The one article I found showed it as being just as performant as statically compiled Java.

<https://ennerf.medium.com/jdk-15-a-world-beyond-nashorn-e19f4258c76a>

**Reply by skbeh** [reply]
The compiling overhead is very big, so it may be not suitable for our usage.
